yoyodyne/models/hmm.py

```diff
@@ -0,0 +1,421 @@
+
+        Simply feeds a BOS string to decoder to provide initial probability.
+        """
+        batch_size, _ = encoder_mask.shape
+        decoder_hiddens = self.init_hiddens(batch_size, self.decoder_layers)
+        bos = (
+            torch.tensor(
+                [self.start_idx], device=self.device, dtype=torch.long
+            )
+            .repeat(batch_size)
+            .unsqueeze(-1)
+        )
+        return self.decode_step(
+            bos, decoder_hiddens, encoder_out, encoder_mask
+        )
+
+    def forward(
+        self,
+        batch: data.PaddedBatch,
+    ) -> torch.Tensor:
+        """Runs the encoder-decoder model.
+
+        The emission tensor returned is of shape 
+        (tgt_len x batch_size x src_len x vocab_size), and the
+        transition tensor returned is of shape
+        (tgt_len x batch_size x src_len x src_len).
+
+        Args:
+            batch (data.PaddedBatch).
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor]: tensosr of emission
+                 and transition probabilities for each transition
+                 state (i.e., target symbol).
+        """
+        encoder_out = self.source_encoder(batch.source).output
+        if self.has_features_encoder:
+            encoder_features_out = self.features_encoder(batch.features).output
+            # Averages to flatten embedding.
+            encoder_features_out = encoder_features_out.sum(
+                dim=1, keepdim=True
+            )
+            # Expands length of source.
+            encoder_features_out = encoder_features_out.expand(
+                -1, encoder_out.shape[1], -1
+            )
+            # Concatenates with average.
+            encoder_out = torch.cat(
+                [encoder_out, encoder_features_out], dim=-1
+            )
+        if self.training:
+            output = self.decode(
+                encoder_out,
+                batch.source.mask,
+                batch.target.padded,
+            )
+        else:
+            output = self.greedy_decode(
+                encoder_out,
+                batch.source.mask,
+            )
+        return output
+
+    def decode(
+        self,
+        encoder_out: torch.Tensor,
+        encoder_mask: torch.Tensor,
+        target: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        """Decodes a sequence given the encoded input.
+
+        Decodes until all sequences in a batch have reached [EOS] up to
+        length of `target` args.
+
+        Args:
+                encoder_out (torch.Tensor): batch of encoded input symbols.
+                    Shape: (batch_size x src_len x encoder_hidden*num_dir).
+                encoder_mask (torch.Tensor): mask for the batch of encoded
+                    input symbols. Shape: (batch_size x src_len).
+                target (torch.Tensor): target symbols. Decodes up to
+                    `len(target)` symbols.
+
+        Returns:
```

Comment:
Since Python can't return multiple things we say it returns a single thing, a tuple. I suggested how to restyle this for `forward`: can you do that here too?

Reply:
Which forward?